File: rust/src/problems/s229_majority_element_ii.rs
struct Solution;

impl Solution {
    pub fn majority_element(mut nums: Vec<i32>) -> Vec<i32> {
        if nums.len() < 2 {
            return nums;
        }
        nums.sort_unstable();
        let mut result = vec![];
        let max = nums.len() / 3;
        let mut count = 0;
        let mut prefix = nums[0];
        for i in 0..nums.len() {
            if nums[i] == prefix {
                count += 1;
            } else {
                if count > max {
                    result.push(prefix);
                }
                prefix = nums[i];
                count = 1;
            }

            if i == nums.len() - 1 && count > max {
                result.push(prefix);
            }
        }
        result
    }
}

struct Example {
    input: Vec<i32>,
    output: Vec<i32>,
}

#[test]
pub fn test() {
    let examples = vec![
        Example {
            input: vec![2, 2],
            output: vec![2],
        },
        Example {
            input: vec![3, 2, 3],
            output: vec![3],
        },
        Example {
            input: vec![1],
            output: vec![1],
        },
        Example {
            input: vec![1, 2],
            output: vec![1, 2],
        }
    ];
    for example in examples {
        assert_eq!(Solution::majority_element(example.input), example.output);
    }
}

File: typescript/src/problems/s229_majority_element_ii.ts
/**
 * @link https://leetcode.com/problems/majority-element-ii/
 * @param nums
 */
export function majorityElement(nums: number[]): number[] {
  if (nums.length < 2) {
    return nums;
  }
  const map: { [key: number]: number } = {};
  const max = Math.floor(nums.length / 3);
  const result = new Set<number>();
  for (const num of nums) {
    map[num] = ( map[num] ?? 0) + 1;
    if (map[num] > max) {
      result.add(num);
    }
  }
  return [...result];
}

export function majorityElementII(nums: number[]): number[] {
  if (nums.length < 2) {
    return nums;
  }
  nums.sort((a, b) => a - b);
  const max = Math.floor(nums.length / 3);
  const result: number[] = [];
  let prefix = nums[0];
  let count = 0;
  for (let i = 0; i < nums.length; i++) {
    if (nums[i] === prefix) {
      count += 1;
    } else {
      if (count > max) {
        result.push(prefix);
      }
      prefix = nums[i];
      count = 1;
    }
    if (i === nums.length - 1 && count > max) {
      result.push(prefix);
    }
  }
  return result;
}